### ISSUE TITLE
Initial high-level goals and outline

### ISSUE THREAD

**Issue by rly** [rly/linkml-arrays-paper#2]
Goal: A slightly technical perspective piece describing the problem of representing arrays in schema and how array support within LinkML solves the problem

A rough first outline:

1. Background
	- The historical challenges of representing arrays in schema with rich metadata (e.g., linkages to ontologies)
		- Previous work, including HDMF, [CORAL](https://github.com/jmchandonia/CORAL), the [NeXus data format](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4453170/) for X-ray and neutron scattering and muon spectroscopy, netCDF?, OME-NGFF, geospatial data?, JSON schema arrays, [previous approaches in LinkML](https://linkml.io/linkml/howtos/multidimensional-arrays.html)
	    - Many file formats for arrays - binary file, numpy npy/npz, HDF5, Zarr, N5, JSON, CSV/TSV, grib, tiff, fits
	    - Many APIs for working with these formats - numpy, h5py, zarr, xarray, etc.
	- Need a unifying framework across these technologies
	    - What is LinkML
4. Adding array support in LinkML - some technical details
    - Specification of simple and labeled arrays
    - Generation of Pydantic API for arrays
    - Dumpers and loaders from/to various file formats from/to various (python-based) APIs
    - Validation
5. Examples in at least two applications
    - Neurophysiology
    - Bioimaging
    - Environmental / geospatial
6. Discussion
    - Challenges of open data in science
    - Expanding coverage / use cases of LinkML

Potential target journals:
- Nature Biotechnology, as Perspective
- Nature Scientific Data, as Article
- Elife, as Tools and Resources?
- Gigascience
- IEEE Big Data

We would write a second paper on LinkML arrays for NWB/neurophysiology, more for a neuroscience audience.

**Comment by cmungall** [comment]
https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.1.0/schema.md

**Comment by oruebel** [comment]
Just some links to a few other array-based file formats that may be of interest: 
- https://www.xdmf.org/index.php/XDMF_Model_and_Format 
- http://cxidb.org/cxi.html
- https://docs.unidata.ucar.edu/netcdf-c/current/faq.html#ncFAQGeneral

**Comment by sneakers-the-rat** [comment]
OK am back from vacation and ready to rumble. 

made some child issues off this one to start tracking different pieces and make 'threads' for discussing those sub-points

**Comment by sneakers-the-rat** [comment]
Sry have been preoccupied with events on campus, will be returning to this next week. Numpydantic is near a 1.0 except for tests and docs, so its
Numpydantic 1.0 -> new array range generator -> update nwb-linkml to reflect.

The intro to the paper can happen async, but once we have those pieces in place we can do the meat of the results

**Comment by sneakers-the-rat** [comment]
hey every1. i just "officially" released [numpydantic](https://neuromatch.social/@jonny/112499381823948185), so the next step now is to put that in linkml arrays generator (should only take ~a day) and then rework `nwb-linkml` (a ~week) before starting the paper. 

unfortunately, my employer has decided to commit egregious unfair labor practices in the form of police violence against my students and colleagues, so starting next week I will be on strike and not doing any work that brings any benefit to my employer - and unfortunately my academic work is decidedly within the scope of struck work. The strike will last as long as June 30th when the grad student contracts expire, but may end sooner than that, pretty unclear at this point. hope u all understand

https://www.uaw4811.org/2024-ulp-charges

**Comment by sneakers-the-rat** [comment]
back at work and ready to roll. got numpydantic linkml PR open https://github.com/linkml/linkml/pull/2178

and about to update `nwb-linkml` to reflect all the work we've done with linkml arrays. my goal is to get read and write working with HDF5 and zarr, and demonstrate a `yaml`-based NWB form like

```yaml
my_dataset: !core.nwbfile.NWBFile
  # ... various metadata fields
  acquisition:
    probe_0_lfp: !core.ecephys.LFP
      probe_0_lfp_data: !core.ecephys.ElectricalSeries
        data:
          # ... metadata attributes
          array: # specify using a relative path and hash
            path: probe_0_lfp_data.zarr
            hash:
              value: # some long hash
              type: blake2b
          electrodes: 
            table: !reference /general/extracellular_ephys/electrodes
            array: [0, 1, 2, 3, 4, 5] # inline arrays should work the same as path references
  general:
    extracellular_ephys:
      electrodes:
        # ... and so on
  stimulus:
    presentation:
      my_stimulus_video: !core.base.TimeSeries
        data:
          array: 
            path: my_video.mp4  # videos behave the same as arrays
            hash: # ...
```

in some serialized form, just as an example. which is *sort of* like what's going on in `hdmf_zarr` except it's human readable, it can accept any kind of file (going to add an additional metadata field for plugins needed to read/write files), and it doesn't become dependent on hierarchical folder structure (that can still be used if it's helpful, but in general I think relying on directory structure in formats is another point of coupling between the abstract specification and the concrete implementation that seems to put a low cap on their expressiveness).

**Comment by sneakers-the-rat** [comment]
sneek preview of the schema models https://github.com/p2p-ld/nwb-linkml/tree/linkml-arrays/nwb_linkml/src/nwb_linkml/models/pydantic/core
and the generated schema: https://github.com/p2p-ld/nwb-linkml/tree/linkml-arrays/nwb_linkml/src/nwb_linkml/schema/linkml/core

that was super easy. need to make some changes in upstream linkml and i think i can scrap pretty much all of my monkeypatching of the generator too. then it's just a matter of writing dumpers and loaders. 

one of those times where you thank yourself for overengineering something before bc it was extremely simple to just swap out the translation/generation routine here. i think u all are gonna get a kick out of nwb-linkml once it's working by how simple it is

**Comment by sneakers-the-rat** [comment]
Alright my dogs after a long delay, i'm gonna enter into writing mode on this starting wednesday. My goal (!!) is to shoot for 2 week writing sprint to get a draft in place, and i'm going to be doing additional demos and proofs of concept as i go.

What I want to shoot for
- Demo of custom schemas alongside NWB
- FastAPI demo with pydantic models & nwb
- SQL db dump with sqlmodel (which will require an additional generator, but shouldn't be too tricky)

but as always this is just what's going into my draft, and so if y'all want to add other things then by all means. I'm going to be doing my own limited bit of history but I think @cmungall and @oruebel your input on the history of arrays in linked data as well as the need for the hdmf-schema and hdmf would be super valuable here :)

lets get this show on the road !!!!!

**Comment by rly** [comment]
Great! November for me has unexpectedly filled up with grant writing, but I will try to help with writing and reviewing where I can.

**Comment by sneakers-the-rat** [comment]
no problem. reading back the initial issue, i realize as i am drafting that i am mixing in some of the nwb stuff just because it's part of the same thought to me. i figure since we were planning on twinned pieces that we can rearrange and split the words on the page once they are there if we want to, but i am starting with the stuff that would be for the perspective piece re: the schema-centric linkml-arrays approach to a data standard

**Comment by rly** [comment]
@melonora suggested today that we could show an implementation of the array spec in another language, e.g., Javascript, to demonstrate how the array spec is not language-specific. Given that JSON Schema support for arrays is lacking, showcasing Javascript support might add value/interest to the paper. I would say this is a nice to have addition if we have time to implement it.

@melonora also mentioned that people in the OME space were asking about how to use the labeled array spec in Java for use in Fiji.

**Comment by rly** [comment]
@sneakers-the-rat Do you have a rough draft that you could share and we can start assigning sections to write / add to / review?

**Comment by sneakers-the-rat** [comment]
apologies for my absence here - i have been waylaid by a time-sensitive project w.r.t. all the scientific data being memoryholed by the state. any of the minimal writing i have done is in this repo, i think sections split out pretty naturally between us

- ryan/oliver: nwb-schema-language and nwb, the state of arrays from the scientific data standards side. how the "stack" of data standards typically work, the link between schema, code, and serialization.
- chris: linkml and history of e.g. datacube and other linked data attempts at array specs, the challenge of representing ordered/array data in rdf-like universe
- me: evolution of python typing ecosystem and how linkml-arrays can be both arbitrary in schema form and program form, i can also handle glue between these sections.

working on a scipy proposal right now. last minute, i know, but so is everything in my life lately

**Comment by sneakers-the-rat** [comment]
working on this over here: https://git.jon-e.net/jonny/scipy-2025

**Comment by melonora** [comment]
Hi all, just to let you know, Ryan and I started to do some free writing as well. You should all have gotten an invite to collaborate on the google doc by now.

**Comment by rly** [comment]
Target deadline for submission: SciPy conference at beginning of July.
